// @heroui/react/dist/components/tabs/tabs.js
"use client";
import { tabsVariants } from '@heroui/styles';
import React__default, { createContext, use, useRef } from 'react';
import { SelectionIndicator } from 'react-aria-components/SelectionIndicator';
import { Tabs, TabPanel as TabPanel$1, Tab as Tab$1, TabList as TabList$1 } from 'react-aria-components/Tabs';
import { composeTwRenderProps, composeSlotClassName } from '../../utils/compose.js';
import { dom } from '../../utils/dom.js';
import { IconChevronUp, IconChevronLeft, IconChevronDown, IconChevronRight } from '../icons.js';
import { ScrollShadow } from '../scroll-shadow/index.js';
import { jsx, jsxs } from 'react/jsx-runtime';

const TabsContext = /*#__PURE__*/createContext({});

/* -------------------------------------------------------------------------------------------------
 * Tabs Root
 * -----------------------------------------------------------------------------------------------*/

const TabsRoot = ({
  children,
  className,
  orientation = "horizontal",
  variant,
  ...props
}) => {
  const slots = React__default.useMemo(() => tabsVariants({
    variant
  }), [variant]);
  return /*#__PURE__*/jsx(TabsContext, {
    value: {
      orientation,
      slots
    },
    children: /*#__PURE__*/jsx(Tabs, {
      ...props,
      className: composeTwRenderProps(className, slots.base()),
      "data-slot": "tabs",
      orientation: orientation,
      children: children
    })
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tabs List Container
 * -----------------------------------------------------------------------------------------------*/

const TabListContainer = ({
  children,
  className,
  ...props
}) => {
  const {
    orientation = "horizontal",
    slots
  } = use(TabsContext);
  const scrollerRef = useRef(null);
  const isVertical = orientation === "vertical";
  const scrollBy = direction => {
    const el = scrollerRef.current;
    if (!el) return;
    const size = isVertical ? el.clientHeight : el.clientWidth;
    el.scrollBy({
      behavior: "smooth",
      [isVertical ? "top" : "left"]: direction * size * 0.8
    });
  };
  return /*#__PURE__*/jsxs(dom.div, {
    className: composeSlotClassName(slots?.tabListContainer, className),
    "data-slot": "tabs-list-container",
    ...props,
    children: [/*#__PURE__*/jsx(ScrollShadow, {
      ref: scrollerRef,
      hideScrollBar: true,
      className: composeSlotClassName(slots?.scroller),
      orientation: orientation,
      size: 64,
      children: children
    }), /*#__PURE__*/jsx("button", {
      "aria-label": isVertical ? "Scroll tabs up" : "Scroll tabs left",
      className: composeSlotClassName(slots?.scrollPrev),
      tabIndex: -1,
      type: "button",
      onClick: () => scrollBy(-1),
      children: isVertical ? /*#__PURE__*/jsx(IconChevronUp, {}) : /*#__PURE__*/jsx(IconChevronLeft, {})
    }), /*#__PURE__*/jsx("button", {
      "aria-label": isVertical ? "Scroll tabs down" : "Scroll tabs right",
      className: composeSlotClassName(slots?.scrollNext),
      tabIndex: -1,
      type: "button",
      onClick: () => scrollBy(1),
      children: isVertical ? /*#__PURE__*/jsx(IconChevronDown, {}) : /*#__PURE__*/jsx(IconChevronRight, {})
    })]
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tabs List
 * -----------------------------------------------------------------------------------------------*/

const TabList = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = use(TabsContext);
  return /*#__PURE__*/jsx(TabList$1, {
    ...props,
    className: composeTwRenderProps(className, slots?.tabList()),
    "data-slot": "tabs-list",
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tab
 * -----------------------------------------------------------------------------------------------*/

const Tab = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = use(TabsContext);
  return /*#__PURE__*/jsx(Tab$1, {
    ...props,
    className: composeTwRenderProps(className, slots?.tab()),
    "data-slot": "tabs-tab",
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tab Indicator
 * -----------------------------------------------------------------------------------------------*/

const TabIndicator = ({
  className,
  ...props
}) => {
  const {
    slots
  } = use(TabsContext);
  return /*#__PURE__*/jsx(SelectionIndicator, {
    className: composeSlotClassName(slots?.tabIndicator, className),
    "data-slot": "tabs-indicator",
    ...props
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tab Panel
 * -----------------------------------------------------------------------------------------------*/

const TabPanel = ({
  children,
  className,
  ...props
}) => {
  const {
    slots
  } = use(TabsContext);
  return /*#__PURE__*/jsx(TabPanel$1, {
    ...props,
    className: composeTwRenderProps(className, slots?.tabPanel()),
    "data-slot": "tabs-panel",
    children: children
  });
};

/* -------------------------------------------------------------------------------------------------
 * Tab Separator
 * -----------------------------------------------------------------------------------------------*/

const TabSeparator = ({
  className,
  ...props
}) => {
  const {
    slots
  } = use(TabsContext);
  return /*#__PURE__*/jsx(dom.span, {
    "aria-hidden": "true",
    className: composeSlotClassName(slots?.separator, className),
    "data-slot": "tabs-separator",
    ...props
  });
};

export { Tab, TabIndicator, TabList, TabListContainer, TabPanel, TabSeparator, TabsRoot };
